numpy/core/tests/test_scalarbuffer.py

```diff
@@ -0,0 +1,96 @@
+"""
+Test scalar buffer interface adheres to PEP 3118
+"""
+import sys
+import numpy as np
+from numpy.testing import run_module_suite, assert_, dec
+
+# types
+scalars = [
+    np.bool_, np.bool8, np.byte, np.short, np.intc, np.int_,
+    np.longlong, np.intp, np.int8, np.int16, np.int32, np.int64,
+    np.ubyte, np.ushort, np.uintc, np.uint, np.ulonglong,
+    np.uintp, np.uint8, np.uint16, np.uint32, np.uint64,
+    np.half, np.single, np.double, np.float_, np.longfloat,
+    np.float16, np.float32, np.float64, np.csingle, np.complex_,
+    np.clongfloat, np.complex64, np.complex128,
+]
+
+scalars_set_code = [
+    (np.bool_, '?'),
+    (np.byte, 'b'),
+    (np.short, 'h'),
+    (np.intc, 'i'),
+    (np.int_, 'l'),
+    (np.longlong, 'q'),
+    (np.ubyte, 'B'),
+    (np.ushort, 'H'),
+    (np.uintc, 'I'),
+    (np.uint, 'L'),
+    (np.ulonglong, 'Q'),
+    (np.half, 'e'),
+    (np.single, 'f'),
+    (np.double, 'd'),
+    (np.float_, 'd'),
+    (np.longfloat, 'g'),
+    (np.csingle, 'Zf'),
+    (np.complex_, 'Zd'),
+    (np.clongfloat, 'Zg'),
+]
+
+# platform dependant dtypes
+for dtype in ('float96', 'float128', 'complex192', 'complex256'):
+    if hasattr(np, dtype):
+        scalars.append(getattr(np, dtype))
+
+
+class TestScalarPEP3118(object):
+    @dec.skipif(sys.version_info.major < 3, "scalars do not implement buffer interface in Python 2")
+    def test_scalar_match_array(self):
+        for scalar in scalars:
+            x = scalar()
+            a = np.array([], dtype=np.dtype(scalar))
+            mv_x = memoryview(x)
+            mv_a = memoryview(a)
+            assert_(mv_x.format == mv_a.format)
+
+    @dec.skipif(sys.version_info.major < 3, "scalars do not implement buffer interface in Python 2")
+    def test_scalar_dim(self):
+        for scalar in scalars:
+            x = scalar()
+            mv_x = memoryview(x)
+            assert_(mv_x.itemsize == np.dtype(scalar).itemsize)
+            assert_(mv_x.ndim == 0)
+            assert_(mv_x.shape == ())
+            assert_(mv_x.strides == ())
+            assert_(mv_x.suboffsets == ())
+
+    @dec.skipif(sys.version_info.major < 3, "scalars do not implement buffer interface in Python 2")
+    def test_scalar_known_code(self):
+        for scalar, code in scalars_set_code:
+            x = scalar()
+            mv_x = memoryview(x)
+            assert_(mv_x.format == code)
+
+    @dec.skipif(sys.version_info.major < 3, "scalars do not implement buffer interface in Python 2")
+    def test_void_scalar_structured_data(self):
+        dt = np.dtype([('name', np.unicode_, 16), ('grades', np.float64, (2,))])
+        a = np.array([('Sarah', (8.0, 7.0)), ('John', (6.0, 7.0))], dtype=dt)
```

Comment:
You might want to consider making this an 0d array, as `np.array(('Sarah', (8.0, 7.0)), ('John', (6.0, 7.0)), dtype=dt)`

Reply:
Not sure I follow, how would a 0d array help?

I'm trying to get an `np.void` object to test. Ideally I would directly make an `np.void` scalar, but I don't see how. 

`np.array(('Sarah', (8.0, 7.0)), dtype=dt)` ndim is 0, but type is ndarray

Reply:
You can convert an 0d array to a scalar with `[()]`, in place of [0]`.

You could then test that the buffer returned from array vs scalar has the same strides, shape, and format

Reply:
I would prefer to just change to:
```python
x = np.array(('ndarray_scalar', (1.2, 3.0)), dtype=dt)[()]
```
This way
  * `x` is definitely different than `a`, not just an element of that array.
  * Its as close to directly creating a structured `np.void` as I can get.
  * ndim/shape/strides/suboffsets are tested against explicit values per PEP3118, not just that they match corresponding buffer fields from ndarray `a`

As a trade off, the format string of the scalar is tested against ndarray format string. Its simpler then testing against the multitude of different, but still valid, PEP3118 format strings for this structure.